src/redisearch_rs/c_entrypoint/inverted_index_ffi/src/lib.rs

```diff
@@ -479,3 +503,465 @@ pub unsafe extern "C" fn InvertedIndex_LastId(ii: *const InvertedIndex) -> t_doc
+/// must be freed using [`IndexReader_Free`] when no longer needed.
+///
+/// # Safety
+///
+/// The following invariant must be upheld when calling this function:
+/// - `ii` must be a valid, non NULL, pointer to an `InvertedIndex` instance.
+///
+/// # Panics
+/// This function will panic if the provided filter is not compatible with the `InvertedIndex` type.
+#[allow(improper_ctypes_definitions)] // `ctx` might contain `t_fieldMask` which is a `u128` type
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn NewIndexReader(
+    ii: *const InvertedIndex,
+    ctx: ReadFilter,
+) -> *mut IndexReader {
+    debug_assert!(!ii.is_null(), "ii must not be null");
+
+    // SAFETY: The caller must ensure that `ii` is a valid pointer to an `InvertedIndex`
+    let ii = unsafe { &*ii };
+
+    let reader = match (ii, ctx) {
+        (InvertedIndex::Full(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::Full(ii.reader(mask))
+        }
+        (InvertedIndex::FullWide(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FullWide(ii.reader(mask))
+        }
+        (InvertedIndex::FreqsFields(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FreqsFields(ii.reader(mask))
+        }
+        (InvertedIndex::FreqsFieldsWide(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FreqsFieldsWide(ii.reader(mask))
+        }
+        (InvertedIndex::FreqsOnly(ii), _) => IndexReader::FreqsOnly(ii.reader()),
+        (InvertedIndex::FieldsOnly(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FieldsOnly(ii.reader(mask))
+        }
+        (InvertedIndex::FieldsOnlyWide(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FieldsOnlyWide(ii.reader(mask))
+        }
+        (InvertedIndex::FieldsOffsets(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FieldsOffsets(ii.reader(mask))
+        }
+        (InvertedIndex::FieldsOffsetsWide(ii), ReadFilter::FieldMask(mask)) => {
+            IndexReader::FieldsOffsetsWide(ii.reader(mask))
+        }
+        (InvertedIndex::OffsetsOnly(ii), _) => IndexReader::OffsetsOnly(ii.reader()),
+        (InvertedIndex::FreqsOffsets(ii), _) => IndexReader::FreqsOffsets(ii.reader()),
+        (InvertedIndex::DocumentIdOnly(ii), _) => IndexReader::DocumentIdOnly(ii.reader()),
+        (InvertedIndex::RawDocumentIdOnly(ii), _) => IndexReader::RawDocumentIdOnly(ii.reader()),
+        (InvertedIndex::Numeric(ii), ReadFilter::None) => IndexReader::Numeric(ii.reader()),
+        (InvertedIndex::Numeric(ii), ReadFilter::Numeric(filter)) if filter.is_numeric_filter() => {
+            IndexReader::NumericFiltered(FilterNumericReader::new(filter, ii.reader()))
+        }
+        (InvertedIndex::Numeric(ii), ReadFilter::Numeric(filter)) => {
+            IndexReader::NumericGeoFiltered(FilterGeoReader::new(filter, ii.reader()))
+        }
+        (index, filter) => panic!("Unsupported filter ({filter:?}) for inverted index ({index})"),
```

Comment:
why is this panic acceptable and is there no way we can prevent this within the type system?

Reply:
I explained the panic with a comment in https://github.com/RediSearch/RediSearch/pull/6837/commits/d14e55429a14657f48531206db27e11a3cbb7f67